src/Compilers/CSharp/Portable/Binder/BindingDiagnosticBag.cs

```diff
@@ -0,0 +1,165 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+#nullable enable
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    internal static class BindingDiagnosticBagExtensions
+    {
+        internal static CSDiagnosticInfo Add(this CodeAnalysis.BindingDiagnosticBag diagnostics, ErrorCode code, Location location)
+        {
+            var info = new CSDiagnosticInfo(code);
+            diagnostics.DiagnosticBag?.Add(new CSDiagnostic(info, location));
+            return info;
+        }
+
+        internal static CSDiagnosticInfo Add(this CodeAnalysis.BindingDiagnosticBag diagnostics, ErrorCode code, Location location, params object[] args)
+        {
+            var info = new CSDiagnosticInfo(code, args);
+            diagnostics.DiagnosticBag?.Add(new CSDiagnostic(info, location));
+            return info;
+        }
+
+        internal static CSDiagnosticInfo Add(this CodeAnalysis.BindingDiagnosticBag diagnostics, ErrorCode code, Location location, ImmutableArray<Symbol> symbols, params object[] args)
+        {
+            var info = new CSDiagnosticInfo(code, args, symbols, ImmutableArray<global::Microsoft.CodeAnalysis.Location>.Empty);
+            diagnostics.DiagnosticBag?.Add(new CSDiagnostic(info, location));
+            return info;
+        }
+
+        internal static void Add(this CodeAnalysis.BindingDiagnosticBag diagnostics, DiagnosticInfo? info, Location location)
+        {
+            if (info is object)
+            {
+                diagnostics.DiagnosticBag?.Add(info, location);
+            }
+        }
+
+        internal static void AddForSymbol(this ref CompoundUseSiteInfo<AssemblySymbol> useSiteInfo, Symbol? symbol, bool addDiagnostics = true, bool addDependencies = true)
+        {
+            if (symbol is null || useSiteInfo.IsDiscarded || !(addDiagnostics || addDependencies))
+            {
+                return;
+            }
+
+            var info = symbol.GetUseSiteInfo();
+
+            if (addDiagnostics)
+            {
+                useSiteInfo.AddDiagnostics(info);
+            }
+            else
+            {
+                Debug.Assert(info.DiagnosticInfo == null || addDependencies); // It would be strange to drop diagnostics, but record dependencies
+            }
+
+            if (addDependencies)
+            {
+                useSiteInfo.AddDependencies(info);
+            }
+        }
+    }
+
+    internal sealed class BindingDiagnosticBag : BindingDiagnosticBag<AssemblySymbol>
```

Comment:
>BindingDiagnosticBag [](start = 26, length = 20)

It is super confusing that we have a non-generic class `BindingDiagnosticBag` declared in `Microsoft.CodeAnalysis` and one declared here in `Microsoft.CodeAnalysis.CSharp`.  I don't know how I'm supposed to keep track of which is which or what the relationship is between them.  Please rename one or both of them, and add some documentation explaining what's what.

Reply:
> It is super confusing that we have a non-generic class BindingDiagnosticBag declared in Microsoft.CodeAnalysis and one declared here in Microsoft.CodeAnalysis.CSharp.  I don't know how I'm supposed to keep track of which is which or what the relationship is between them.  Please rename one or both of them, and add some documentation explaining what's what.

If you think of better names, please suggests them. 
```Microsoft.CodeAnalysis.BindingDiagnosticBag``` is the root base class for all BindingDiagnosticBags. It is used in a shared structures and shared code related to attribute binding, that is why having a non-generic base is convenient. In all other places we are dealing with language specific ```Microsoft.CodeAnalysis.CSharp.BindingDiagnosticBag``` or ```Microsoft.CodeAnalysis.VisualBasic.BindingDiagnosticBag```, which have an intermediate base ```Microsoft.CodeAnalysis.CSharp.BindingDiagnosticBag`1``` with shared functionality.

I don't think you absolutely have to keep track of which BindingDiagnosticBag is which because the inheritance hierarchy ensures that APIs do the right thing, or not available.

---
In reply to: [361553212](https://github.com/dotnet/roslyn/pull/40560#discussion_r361553212) [](ancestors = 361553212)

Reply:
BTW, I am in the process of refactoring VB and removing temporary conversion operators added to the root base class. I am also marking that base class abstract. So, only leaf, language specific, classes can be instantiated. Whenever you see a "BindingDiagnosticBag", you can be certain you are dealing with an instance of a leaf class, defined either in VB or in C# compiler. 

---
In reply to: [361555234](https://github.com/dotnet/roslyn/pull/40560#discussion_r361555234) [](ancestors = 361555234,361553212)

Reply:
I think the root of the hierarchy could have `Base` added to its name to help the reader understand (in a separate PR)

---
In reply to: [361692792](https://github.com/dotnet/roslyn/pull/40560#discussion_r361692792) [](ancestors = 361692792,361555234,361553212)

Reply:
> I think the root of the hierarchy could have Base added to its name to help the reader understand (in a separate PR)

Added a PROTOTYPE comment in [https://github.com/dotnet/roslyn/pull/40640](https://github.com/dotnet/roslyn/pull/40640)

---
In reply to: [361746115](https://github.com/dotnet/roslyn/pull/40560#discussion_r361746115) [](ancestors = 361746115,361692792,361555234,361553212)